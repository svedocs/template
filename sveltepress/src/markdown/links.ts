import type { Plugin } from 'unified'
import { visit } from 'unist-util-visit'

// parse [foo](bar) to <Link label="foo" to="bar" />
const anchors: Plugin<any[], any> = () => {
  return (tree) => {
    visit(tree, (node, idx, parent) => {
      if (node.type === 'link') {
        parent.children.splice(idx, 1, {
          type: 'html',
          value: `<Link to="${node.url}">{#snippet labelRenderer()}`,
        }, ...node.children, {
          type: 'html',
          value: '{/snippet}</Link>',
        }, {
          type: 'SvpHeading',
          value: node.children[0].value,
        })
      }
      if (node.type === 'SvpHeading') {
        parent.children.splice(idx, 1)
      }
    })
  }
}

export default anchors
