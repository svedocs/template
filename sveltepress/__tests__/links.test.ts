import { mdToSvelte } from '@sveltepress/vite'
import { describe, expect, it } from 'vitest'
import links from '../src/markdown/links'

const md = `
## Foo

[foo](https://www.google.com/)
[bar](/foo/bar)
[\`with inline code\`](/foo/bar)

`

describe('links', () => {
  it('simple', async () => {
    const r = await mdToSvelte({
      mdContent: md,
      filename: 'demo.md',
      remarkPlugins: [links, () => (tree) => {
        expect(JSON.stringify(tree, null, 2)).toMatchSnapshot()
      }],
    })
    expect(r).toMatchInlineSnapshot(`
      {
        "code": "<h2>Foo</h2>
      <p><Link to="https://www.google.com/">{#snippet labelRenderer()}foo{/snippet}</Link>
      <Link to="/foo/bar">{#snippet labelRenderer()}bar{/snippet}</Link>
      <Link to="/foo/bar">{#snippet labelRenderer()}<code>{\`with inline code\`}</code>{/snippet}</Link></p>",
        "data": {},
      }
    `)
  })
})
