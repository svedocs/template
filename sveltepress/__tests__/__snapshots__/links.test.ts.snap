// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`links > simple 1`] = `
"{
  "type": "root",
  "children": [
    {
      "type": "heading",
      "depth": 2,
      "children": [
        {
          "type": "text",
          "value": "Foo",
          "position": {
            "start": {
              "line": 2,
              "column": 4,
              "offset": 4
            },
            "end": {
              "line": 2,
              "column": 7,
              "offset": 7
            }
          }
        }
      ],
      "position": {
        "start": {
          "line": 2,
          "column": 1,
          "offset": 1
        },
        "end": {
          "line": 2,
          "column": 7,
          "offset": 7
        }
      }
    },
    {
      "type": "paragraph",
      "children": [
        {
          "type": "html",
          "value": "<Link to=\\"https://www.google.com/\\">{#snippet labelRenderer()}"
        },
        {
          "type": "text",
          "value": "foo",
          "position": {
            "start": {
              "line": 4,
              "column": 2,
              "offset": 10
            },
            "end": {
              "line": 4,
              "column": 5,
              "offset": 13
            }
          }
        },
        {
          "type": "html",
          "value": "{/snippet}</Link>"
        },
        {
          "type": "text",
          "value": "\\n",
          "position": {
            "start": {
              "line": 4,
              "column": 31,
              "offset": 39
            },
            "end": {
              "line": 5,
              "column": 1,
              "offset": 40
            }
          }
        },
        {
          "type": "html",
          "value": "<Link to=\\"/foo/bar\\">{#snippet labelRenderer()}"
        },
        {
          "type": "text",
          "value": "bar",
          "position": {
            "start": {
              "line": 5,
              "column": 2,
              "offset": 41
            },
            "end": {
              "line": 5,
              "column": 5,
              "offset": 44
            }
          }
        },
        {
          "type": "html",
          "value": "{/snippet}</Link>"
        },
        {
          "type": "text",
          "value": "\\n",
          "position": {
            "start": {
              "line": 5,
              "column": 16,
              "offset": 55
            },
            "end": {
              "line": 6,
              "column": 1,
              "offset": 56
            }
          }
        },
        {
          "type": "html",
          "value": "<Link to=\\"/foo/bar\\">{#snippet labelRenderer()}"
        },
        {
          "type": "inlineCode",
          "value": "with inline code",
          "position": {
            "start": {
              "line": 6,
              "column": 2,
              "offset": 57
            },
            "end": {
              "line": 6,
              "column": 20,
              "offset": 75
            }
          }
        },
        {
          "type": "html",
          "value": "{/snippet}</Link>"
        }
      ],
      "position": {
        "start": {
          "line": 4,
          "column": 1,
          "offset": 9
        },
        "end": {
          "line": 6,
          "column": 31,
          "offset": 86
        }
      }
    }
  ],
  "position": {
    "start": {
      "line": 1,
      "column": 1,
      "offset": 0
    },
    "end": {
      "line": 8,
      "column": 1,
      "offset": 88
    }
  }
}"
`;
